autoregressive solver

Davide-Miotti opened the pull request:
## Description
Implements  `AutoregressiveSolver`.
This is a tentative initial implementation, not a definive development.

### `AutoregressiveSolver`
1.  Takes `input` ($y_t$).
2.  Predicts $\hat{y}_{t+1}$.
3.  Calculates loss against `unroll` target ($y_{t+1}$).
4.  Weights the loss using the condition's strategy.
5.  **Feeds $\hat{y}_{t+1}$ back as input** for the next step.
6.  Repeats for `unroll_length`.

### Limitations
* **Single length input** does not remember previous states beyond the immediate last prediction, limiting context for long-term dependencies.
* **For curriculum learning** you need to define multiple conditions with different unroll length by hand, no automation

This PR fixes #730 

## Checklist

- [x] Code follows the project’s [Code Style Guidelines](https://github.com/mathLab/PINA/blob/master/CONTRIBUTING.md#code-style--guidelines)
- [x] Tests have been added or updated
- [ ] Documentation has been updated if necessary
- [x] Pull request is linked to an open issue

Davide-Miotti commented:
To give some additional context, here is a python script that uses the code from this PR:
[autoregressive_prova.py](https://github.com/user-attachments/files/24171475/autoregressive_prova.py)

Davide-Miotti commented:
With the last commit I deleted the autoregressive condition and now all the logic is contained within autoregressive solver, I also pushed by mistake a test script, called `autoregressive_prova_generic_condition.py` which can be used to test the new pipeline. I'll live it there for you to try and remove it later before eventual merge.
Now the solver is set to use `DataCondition`, as was suggested by @dario-coscia and adopts an adaptive weighting for unrolls as suggested by @GiovanniCanali
Side note, the automatic random shuffling performed by the dataloader is disruptive for autoregressive method, in the test script I set it to false in the `Trainer` settings

FilippoOlivo commented:
Hi @GiovanniCanali @Davide-Miotti, since this PR is related to 0.3 release, I changed the base branch

Davide-Miotti commented:
With the last edits I moved the whole logic to `AutoregressiveSolver` in `pina/solver/autoregressive_solver/autoregressive_solver.py` and to `AutoregressiveSolverInterface` in `pina/solver/autoregressive_solver/autoregressive_solver_interface.py`. The logic for splitting input data into multiple unrolls is implemented through a `@dataclass UnrollInstructions` defined in the same file as `AutoregressiveSolverInterface`. 

This solver does not support subdivision into training and validation data. I believe we should wait for latest version of conditions and data loader for that one, because any random splitting of the data breaks the temporal sequence required by the autoregressive solver.

dario-coscia commented on pina/solver/autoregressive_solver/autoregressive_solver_interface.py:
```diff
@@ -0,0 +1,157 @@
+        pass
+
+    @abstractmethod
+    def predict(self, initial_state, num_steps):
```
maybe  `unroll`?

Davide-Miotti replied:
called it predict because it is the function you call once the model is trained, `unroll` is equally good

dario-coscia commented on pina/solver/autoregressive_solver/autoregressive_solver.py:
```diff
@@ -0,0 +1,225 @@
+        :rtype: torch.Tensor
+        """
+
+        initial_data, unroll_data = self.create_unroll_windows(
```
be aware that you are unrolling for each epoch and each conditon, this is very memory expensive

dario-coscia commented on pina/solver/autoregressive_solver/autoregressive_solver.py:
```diff
@@ -0,0 +1,225 @@
+
+        return (step_losses * weights).sum()
+
+    def create_unroll_windows(
```
we should do it at beginning of training, we can use [`on_train_start`](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.hooks.ModelHooks.html#lightning.pytorch.core.hooks.ModelHooks.on_train_start)

Davide-Miotti replied:
I implemented a separate function, that can be called when the problem is defined, see attached test script

dario-coscia commented on pina/solver/autoregressive_solver/autoregressive_solver.py:
```diff
@@ -0,0 +1,225 @@
+        )  # [num_unrolls, unroll_length, features]
+        return initial_data, unroll_data
+
+    def decide_starting_indices(
```
we should do it at beginning of training, we can use [`on_train_start`](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.hooks.ModelHooks.html#lightning.pytorch.core.hooks.ModelHooks.on_train_start)

Davide-Miotti replied:
I made it a `@staticmethod` that can be called outside as well

Davide-Miotti commented:
I am preparing a test script that show how it all works, will load it in the next days
[autoregressive_codice_prova.py](https://github.com/user-attachments/files/24648969/autoregressive_codice_prova.py)

Davide-Miotti commented:
Implemented the changes and instead of preparing a separate script, I added an end-to-end test that shows how the solver can be used.
Also corrected the docstrings, do I need to also update the documentation somewhere else?

GiovanniCanali commented:
> Implemented the changes and instead of preparing a separate script, I added an end-to-end test that shows how the solver can be used. Also corrected the docstrings, do I need to also update the documentation somewhere else?

Thank you @Davide-Miotti!

Yes, you also need to add the .rst files under `PINA/docs/source/_rst`. In particular, please update `_code.rst` accordingly and add a `<file_name>.rst` file for each new file you introduced. You can refer to the existing files in that directory as a reference.

GiovanniCanali commented:
Memo: move the `unroll` logic and the `settings` dictionary workaround to the `AutoregressiveCondition` once available.

dario-coscia commented:
Hey! Very good job! Since I see many commit messages can you squash in two/three main commits?

GiovanniCanali commented:
> Hey! Very good job! Since I see many commit messages can you squash in two/three main commits?

Hi @dario-coscia,
We will squash the commits before merging. Since this is still work in progress, we prefer to keep the history as is for now to avoid losing any relevant changes.
